pkg/utils/k8sutil.go

```diff
@@ -265,247 +268,98 @@ func GetFunctionData(runtime, ftype, modName string) (imageName, depName, fileNa
 	return
 }
 
+type metadata struct {
+	Name       string
+	Namespace  string
+	Handler    string
+	Module     string
+	Image      string
+	File       string
+	Dependency string
+	Labels     map[string]string
+}
+
 // EnsureK8sResources creates/updates k8s objects (deploy, svc, configmap) for the function
 func EnsureK8sResources(ns, name string, funcObj *spec.Function, client kubernetes.Interface) error {
 	str := strings.Split(funcObj.Spec.Handler, ".")
 	if len(str) != 2 {
 		return errors.New("Failed: incorrect handler format. It should be module_name.handler_name")
 	}
+
 	funcHandler := str[1]
 	modName := str[0]
+
 	fileName := modName
 	imageName, depName, fileName, err := GetFunctionData(funcObj.Spec.Runtime, funcObj.Spec.Type, modName)
-
 	if err != nil {
 		return err
 	}
 
-	//add configmap
 	labels := map[string]string{
 		"function": name,
 	}
 	for k, v := range funcObj.Metadata.Labels {
 		labels[k] = v
 	}
 
+	funcMeta := metadata{
+		Name:       name,
+		Namespace:  ns,
+		Handler:    funcHandler,
+		Module:     modName,
+		Image:      imageName,
+		File:       fileName,
+		Dependency: depName,
+		Labels:     labels,
+	}
+
 	t := true
 	or := []metav1.OwnerReference{
 		{
 			Kind:               "Function",
 			APIVersion:         "k8s.io",
-			Name:               name,
+			Name:               funcMeta.Name,
 			UID:                funcObj.Metadata.UID,
 			BlockOwnerDeletion: &t,
 		},
 	}
 
-	podAnnotations := map[string]string{
-		// Attempt to attract the attention of prometheus.
-		// For runtimes that don't support /metrics,
-		// prometheus will get a 404 and mostly silently
-		// ignore the pod (still displayed in the list of
-		// "targets")
-		"prometheus.io/scrape": "true",
-		"prometheus.io/path":   "/metrics",
-		"prometheus.io/port":   "8080",
-	}
-
-	configMap := &v1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            name,
-			Labels:          labels,
-			OwnerReferences: or,
-		},
-		Data: map[string]string{
-			"handler": funcObj.Spec.Handler,
-			fileName:  funcObj.Spec.Function,
-			depName:   funcObj.Spec.Deps,
-		},
-	}
-
-	_, err = client.Core().ConfigMaps(ns).Create(configMap)
-	if err != nil && k8sErrors.IsAlreadyExists(err) {
-		data, _ := json.Marshal(configMap)
-		_, err = client.Core().ConfigMaps(ns).Patch(configMap.Name, types.StrategicMergePatchType, data)
-	}
+	err = ensureFuncConfigMap(client, funcMeta, funcObj, or)
```

Comment:
this passes almost all the same information twice afaics - in `funcMeta` and `funcObj`, which (apart from waste) introduces confusion if they ever disagree.

Drop `funcMeta` and just use a real `Function` object.  The only reason you needed `funcMeta` here afaics is because you passed the *spec* rather than the full object into this function in the first place - so just fix that.

Reply:
Originally, I wanted to break the EnsureK8SResources() into small pieces including EnsureFuncConfigMap(), EnsureFuncService(), EnsureFuncDeployment() and EnsureFuncJob(). Thus, I need to pass almost the same arguments into those funcs. To do that, I constructed the funcMeta struct.

In fact, funcMeta and funcObj are not almost the same information. funcMeta contains "post-processing" data from funcObj including image runtime, dependency filename, function filename, function name (aka handler), labels (after appending). The reason I included the funcMeta object is that I don't want to have a long-list of arguments passing to the ensureSomething() function. I agree both objects have some overlaps like name or namespace. If we want to pass the original/real function object, then we end up with the duplication of processing its data everywhere. So I _think_ we can either refactor the original function object or keep passing two structs. WDYT ?